rfd/0017-supply-chain-hardening.md

```diff
@@ -0,0 +1,59 @@
+# Supply Chain Hardening
+
+To prevent supply chain attacks take control over all software distribution channels for Teleport and 
+only distribute signed software.
+
+## Why 
+
+Have a cryptographicaly-verifiable audit trail of all software changes and releases 
+and protect all software distribution channels.
+
+## How
+
+### Github
+
+Require [signed commits](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/about-commit-signature-verification) 
+for protected branches `master` and `branch/x.y`.
+
+Turn on commit signing for users and bots following steps 
+[here](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/about-commit-signature-verification)
+
+**Github Mirror**
+
+Self host git repository for Teleport OSS, Enterprise, Cloud and webapps repositories.
+To allow outside collaboration, allow pull requests, forks, but process all commits using CI/CD pipeline 
+and do not use Github as a source of truth.
```

Comment:
If we used self-hosted Gitlab, it looks like we could use SSH certificates issued by Teleport's user CA to authenticate users to the repo (see https://docs.gitlab.com/ee/administration/operations/ssh_certificates.html). I believe BitBucket also supports SSH CAs.

We'd probably still need to use a GPG key to sign commits, but it this would be an extra security step and would also prevent users from needing to upload their own public keys. Granting/revoking access to the repo would become an easier process (given that we can't just use HTTPS authentication because our repos use submodules)

Reply:
We're doing exactly this and it works pretty well, though there are a couple gotchas. Happy to give more details in slack or something more private than github. :)